components/raftstore/src/store/fsm/peer.rs

```diff
@@ -3764,15 +3763,24 @@ where
         self.ctx.router.close(region_id);
         self.fsm.stop();
 
-        if is_initialized
-            && !merged_by_target
-            && meta
-                .region_ranges
-                .remove(&enc_end_key(self.fsm.peer.region()))
-                .is_none()
-        {
-            panic!("{} meta corruption detected", self.fsm.peer.tag);
+        if !merged_by_target {
+            let key = if is_peer_initialized {
+                let region_info = self.fsm.peer.region();
+                Some(enc_end_key(region_info))
+            } else if is_latest_initialized {
+                let latest_region_info = meta.regions.get(&region_id).unwrap();
+                Some(enc_end_key(latest_region_info))
+            } else {
+                None
+            };
+
+            if let Some(delete_key) = key {
+                if meta.region_ranges.remove(&delete_key).is_none() {
+                    panic!("{} meta corruption detected", self.fsm.peer.tag);
```

Comment:
Could panic happen when key is Some(enc_end_key(latest_region_info))?
If not, maybe this panic code can be moved to if is_peer_initialized block.

Reply:
I haven't seen this panic by now, it checks another kind of consistency between the `region_id` and `end_key`,  maybe it's okay to keep the check here.